File: src/languages/go/harness.go
package main

import (
	"container/heap"
	"fmt"
	"math"
	"reflect"
	"sort"
	"strconv"
	"strings"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Node struct {
	Val       int
	Neighbors []*Node
}

type Interval struct {
	Start int
	End   int
}

func MakeInt(v int) *int {
	return &v
}

func ListToLinkedList(arr []int) *ListNode {
	if len(arr) == 0 {
		return nil
	}
	head := &ListNode{Val: arr[0]}
	curr := head
	for i := 1; i < len(arr); i++ {
		curr.Next = &ListNode{Val: arr[i]}
		curr = curr.Next
	}
	return head
}

func LinkedListToList(head *ListNode) []int {
	res := []int{}
	curr := head
	visited := make(map[*ListNode]bool)
	for curr != nil {
		if visited[curr] {
			break
		}
		visited[curr] = true
		res = append(res, curr.Val)
		curr = curr.Next
	}
	return res
}

func MakeCycle(arr []int, pos int) *ListNode {
	head := ListToLinkedList(arr)
	if pos == -1 || head == nil {
		return head
	}
	tail := head
	var target *ListNode
	idx := 0
	for tail != nil {
		if idx == pos {
			target = tail
		}
		if tail.Next == nil {
			break
		}
		tail = tail.Next
		idx++
	}
	if tail != nil && target != nil {
		tail.Next = target
	}
	return head
}

func IntsToTree(vals ...int) *TreeNode {
	if len(vals) == 0 {
		return nil
	}
	root := &TreeNode{Val: vals[0]}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(vals) {
		node := queue[0]
		queue = queue[1:]
		if i < len(vals) {
			node.Left = &TreeNode{Val: vals[i]}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(vals) {
			node.Right = &TreeNode{Val: vals[i]}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func TreeToInts(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	res := []int{}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		if node != nil {
			res = append(res, node.Val)
			queue = append(queue, node.Left)
			queue = append(queue, node.Right)
		}
	}
	return res
}

func ListToTree(arr []*int) *TreeNode {
	if len(arr) == 0 || arr[0] == nil {
		return nil
	}
	root := &TreeNode{Val: *arr[0]}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(arr) {
		node := queue[0]
		queue = queue[1:]
		if i < len(arr) && arr[i] != nil {
			node.Left = &TreeNode{Val: *arr[i]}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(arr) && arr[i] != nil {
			node.Right = &TreeNode{Val: *arr[i]}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func TreeToList(root *TreeNode) []*int {
	if root == nil {
		return []*int{}
	}
	res := []*int{}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		if node != nil {
			val := node.Val
			res = append(res, &val)
			queue = append(queue, node.Left)
			queue = append(queue, node.Right)
		} else {
			res = append(res, nil)
		}
	}
	for len(res) > 0 && res[len(res)-1] == nil {
		res = res[:len(res)-1]
	}
	return res
}

func BuildGraph(adj [][]int) *Node {
	if len(adj) == 0 {
		return nil
	}
	nodes := make([]*Node, len(adj))
	for i := range adj {
		nodes[i] = &Node{Val: i + 1}
	}
	for i, neighbors := range adj {
		for _, nei := range neighbors {
			nodes[i].Neighbors = append(nodes[i].Neighbors, nodes[nei-1])
		}
	}
	return nodes[0]
}

func GraphToAdj(node *Node) [][]int {
	if node == nil {
		return [][]int{}
	}
	visited := make(map[int]*Node)
	var dfs func(n *Node)
	dfs = func(n *Node) {
		if _, ok := visited[n.Val]; ok {
			return
		}
		visited[n.Val] = n
		for _, nei := range n.Neighbors {
			dfs(nei)
		}
	}
	dfs(node)
	adj := make([][]int, len(visited))
	for i := 1; i <= len(visited); i++ {
		if n, ok := visited[i]; ok {
			row := []int{}
			for _, nei := range n.Neighbors {
				row = append(row, nei.Val)
			}
			adj[i-1] = row
		} else {
			adj[i-1] = []int{}
		}
	}
	return adj
}

func NormalizeNested(groups [][]int) [][]int {
	res := make([][]int, len(groups))
	for i, g := range groups {
		cp := append([]int{}, g...)
		sort.Ints(cp)
		res[i] = cp
	}
	sort.Slice(res, func(i, j int) bool {
		if len(res[i]) != len(res[j]) {
			return len(res[i]) < len(res[j])
		}
		for k := 0; k < len(res[i]); k++ {
			if res[i][k] != res[j][k] {
				return res[i][k] < res[j][k]
			}
		}
		return false
	})
	return res
}

func NormalizeNestedStrings(groups [][]string) [][]string {
	res := make([][]string, len(groups))
	for i, g := range groups {
		cp := append([]string{}, g...)
		sort.Strings(cp)
		res[i] = cp
	}
	sort.Slice(res, func(i, j int) bool {
		if len(res[i]) != len(res[j]) {
			return len(res[i]) < len(res[j])
		}
		for k := 0; k < len(res[i]); k++ {
			if res[i][k] != res[j][k] {
				return res[i][k] < res[j][k]
			}
		}
		return false
	})
	return res
}

func SortStrings(words []string) []string {
	cp := append([]string{}, words...)
	sort.Strings(cp)
	return cp
}

func SortInts(arr []int) []int {
	cp := append([]int{}, arr...)
	sort.Ints(cp)
	return cp
}

func min(a int, rest ...int) int {
	m := a
	for _, v := range rest {
		if v < m {
			m = v
		}
	}
	return m
}

func max(a int, rest ...int) int {
	m := a
	for _, v := range rest {
		if v > m {
			m = v
		}
	}
	return m
}

func deepEqual(expected, actual interface{}) bool {
	if reflect.DeepEqual(expected, actual) {
		return true
	}

	v1 := reflect.ValueOf(expected)
	v2 := reflect.ValueOf(actual)

	if !v1.IsValid() && !v2.IsValid() {
		return true
	}
	if !v1.IsValid() || !v2.IsValid() {
		return false
	}

	// Compare slices / arrays (treating nil slice and empty slice as equal)
	if (v1.Kind() == reflect.Slice || v1.Kind() == reflect.Array) &&
		(v2.Kind() == reflect.Slice || v2.Kind() == reflect.Array) {
		if v1.Len() == 0 && v2.Len() == 0 {
			return true
		}
		if v1.Len() != v2.Len() {
			return false
		}
		for i := 0; i < v1.Len(); i++ {
			if !deepEqual(v1.Index(i).Interface(), v2.Index(i).Interface()) {
				return false
			}
		}
		return true
	}

	// Compare maps (treating nil map and empty map as equal)
	if v1.Kind() == reflect.Map && v2.Kind() == reflect.Map {
		if v1.Len() == 0 && v2.Len() == 0 {
			return true
		}
		if v1.Len() != v2.Len() {
			return false
		}
		for _, key := range v1.MapKeys() {
			val1 := v1.MapIndex(key)
			val2 := v2.MapIndex(key)
			if !val2.IsValid() || !deepEqual(val1.Interface(), val2.Interface()) {
				return false
			}
		}
		return true
	}

	// Compare pointers
	if v1.Kind() == reflect.Ptr && v2.Kind() == reflect.Ptr {
		if v1.IsNil() && v2.IsNil() {
			return true
		}
		if v1.IsNil() || v2.IsNil() {
			return false
		}
		return deepEqual(v1.Elem().Interface(), v2.Elem().Interface())
	}

	// Compare structs of matching type
	if v1.Kind() == reflect.Struct && v2.Kind() == reflect.Struct && v1.Type() == v2.Type() {
		for i := 0; i < v1.NumField(); i++ {
			if !deepEqual(v1.Field(i).Interface(), v2.Field(i).Interface()) {
				return false
			}
		}
		return true
	}

	return false
}

type TestHarness struct{}

var Tests TestHarness

func (t TestHarness) BoolCheck(msg string, b bool) {
	if b {
		fmt.Printf("Test passed: %s\n", msg)
	} else {
		fmt.Printf("Test failed: %s\n", msg)
		panic(fmt.Sprintf("Test failed: %s", msg))
	}
}

func (t TestHarness) EqualCheck(msg string, expected, actual interface{}) {
	if deepEqual(expected, actual) {
		fmt.Printf("Test passed: %s\n", msg)
	} else {
		fmt.Printf("Test failed: %s\nExpected: %#v\nActual:   %#v\n", msg, expected, actual)
		panic(fmt.Sprintf("Test failed: %s", msg))
	}
}

func (t TestHarness) UnorderedEqualCheck(msg string, expected, actual interface{}) {
	switch exp := expected.(type) {
	case [][]int:
		if act, ok := actual.([][]int); ok {
			t.EqualCheck(msg, NormalizeNested(exp), NormalizeNested(act))
			return
		}
	case [][]string:
		if act, ok := actual.([][]string); ok {
			t.EqualCheck(msg, NormalizeNestedStrings(exp), NormalizeNestedStrings(act))
			return
		}
	}
	t.EqualCheck(msg, expected, actual)
}

File: src/languages/ocaml/harness.ml
type list_node = {
  mutable val_ : int;
  mutable next : list_node option;
}

let create_list_node v = { val_ = v; next = None }

let list_to_linked_list (arr : int list) : list_node option =
  match arr with
  | [] -> None
  | hd :: tl ->
      let head = { val_ = hd; next = None } in
      let curr = ref head in
      List.iter (fun x ->
        let n = { val_ = x; next = None } in
        !curr.next <- Some n;
        curr := n
      ) tl;
      Some head

let linked_list_to_list (head : list_node option) : int list =
  let res = ref [] in
  let curr = ref head in
  let count = ref 0 in
  while !curr <> None && !count < 10000 do
    match !curr with
    | None -> ()
    | Some node ->
        res := node.val_ :: !res;
        curr := node.next;
        incr count
  done;
  List.rev !res

let make_cycle (arr : int list) (pos : int) : list_node option =
  let head = list_to_linked_list arr in
  if pos = -1 then head
  else
    match head with
    | None -> None
    | Some h ->
        let target = ref None in
        let tail = ref h in
        let idx = ref 0 in
        let rec loop node i =
          if i = pos then target := Some node;
          match node.next with
          | None -> tail := node
          | Some next_node -> loop next_node (i + 1)
        in
        loop h 0;
        !tail.next <- !target;
        Some h

type tree_node = {
  mutable val_ : int;
  mutable left : tree_node option;
  mutable right : tree_node option;
}

let create_tree_node v = { val_ = v; left = None; right = None }

let list_to_tree (arr : int option list) : tree_node option =
  match arr with
  | [] | None :: _ -> None
  | Some root_val :: rest ->
      let root = { val_ = root_val; left = None; right = None } in
      let q = Queue.create () in
      Queue.push root q;
      let items = ref rest in
      while not (Queue.is_empty q) && !items <> [] do
        let node = Queue.pop q in
        (match !items with
        | [] -> ()
        | l_val :: tl ->
            items := tl;
            (match l_val with
            | None -> ()
            | Some v ->
                let left_node = { val_ = v; left = None; right = None } in
                node.left <- Some left_node;
                Queue.push left_node q));
        (match !items with
        | [] -> ()
        | r_val :: tl ->
            items := tl;
            (match r_val with
            | None -> ()
            | Some v ->
                let right_node = { val_ = v; left = None; right = None } in
                node.right <- Some right_node;
                Queue.push right_node q))
      done;
      Some root

let tree_to_list (root : tree_node option) : int option list =
  match root with
  | None -> []
  | Some r ->
      let q = Queue.create () in
      Queue.push (Some r) q;
      let res = ref [] in
      while not (Queue.is_empty q) do
        let item = Queue.pop q in
        match item with
        | None -> res := None :: !res
        | Some node ->
            res := Some node.val_ :: !res;
            Queue.push node.left q;
            Queue.push node.right q
      done;
      let l = List.rev !res in
      let rec trim_trailing_nones acc = function
        | None :: rest -> trim_trailing_nones acc rest
        | l -> List.rev l
      in
      trim_trailing_nones [] (List.rev l)

let ints_to_tree (arr : int list) : tree_node option =
  list_to_tree (List.map (fun v -> Some v) arr)

let tree_to_ints (root : tree_node option) : int list =
  let raw = tree_to_list root in
  let rec extract = function
    | [] -> []
    | Some v :: rest -> v :: extract rest
    | None :: rest -> extract rest
  in
  extract raw

type graph_node = {
  mutable val_ : int;
  mutable neighbors : graph_node list;
}

let create_graph_node v = { val_ = v; neighbors = [] }

let build_graph (adj : int list list) : graph_node option =
  match adj with
  | [] -> None
  | _ ->
      let n = List.length adj in
      let nodes = Array.init n (fun i -> { val_ = i + 1; neighbors = [] }) in
      let adj_arr = Array.of_list adj in
      Array.iteri (fun i neighbors ->
        nodes.(i).neighbors <- List.map (fun nei -> nodes.(nei - 1)) neighbors
      ) adj_arr;
      Some nodes.(0)

let graph_to_adj (node : graph_node option) : int list list =
  match node with
  | None -> []
  | Some root ->
      let visited = Hashtbl.create 16 in
      let rec dfs n =
        if not (Hashtbl.mem visited n.val_) then begin
          Hashtbl.add visited n.val_ n;
          List.iter dfs n.neighbors
        end
      in
      dfs root;
      let n = Hashtbl.length visited in
      let res = ref [] in
      for i = 1 to n do
        if Hashtbl.mem visited i then
          let gn = Hashtbl.find visited i in
          res := (List.map (fun (nei : graph_node) -> nei.val_) gn.neighbors) :: !res
        else
          res := [] :: !res
      done;
      List.rev !res

type interval = {
  start : int;
  end_ : int;
}

let normalize_nested (groups : 'a list list) : 'a list list =
  List.sort compare (List.map (List.sort compare) groups)

let sort_strings (words : string list) : string list =
  List.sort String.compare words

let sort_ints (arr : int list) : int list =
  List.sort compare arr

module Tests = struct
  let bool_check msg b =
    if b then
      Printf.printf "Test passed: %s\n" msg
    else begin
      Printf.printf "Test failed: %s\n" msg;
      failwith "Test failed"
    end

  let string_check to_str msg expected actual =
    if expected = actual then
      Printf.printf "Test passed: %s\n" msg
    else begin
      Printf.printf "Test failed: %s\nExpected: %s\nActual:   %s\n" msg (to_str expected) (to_str actual);
      failwith "Test failed"
    end

  let equal_check msg expected actual =
    if expected = actual then
      Printf.printf "Test passed: %s\n" msg
    else begin
      Printf.printf "Test failed: %s\n" msg;
      failwith "Test failed"
    end

  let unordered_equal_check msg expected actual =
    let norm_exp = normalize_nested expected in
    let norm_act = normalize_nested actual in
    equal_check msg norm_exp norm_act
end

